File: draft/start_server.c
#include "mongoose.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

static const char *s_http_port = "8000";
static const char *s_web_directory = ".";

static void generate_report() {
    printf("Generating report...\n");
    system("python generate_report.py");
}

static void check_and_generate_report() {
    const char *html_file = "index.html";
    struct stat st;
    time_t current_time = time(NULL);

    if (stat(html_file, &st) == -1 || difftime(current_time, st.st_mtime) > 60) {
        printf("%s is older than 60 seconds or does not exist. Running generate_report.py...\n", html_file);
        generate_report();
    } else {
        printf("%s is up-to-date.\n", html_file);
    }
}

static void serve_text_file(struct mg_connection *c, const char *path) {
    FILE *file = fopen(path, "r");
    if (file == NULL) {
        mg_http_reply(c, 404, NULL, "File not found");
        return;
    }

    fseek(file, 0, SEEK_END);
    long file_size = ftell(file);
    fseek(file, 0, SEEK_SET);

    char *content = malloc(file_size + 1);
    fread(content, 1, file_size, file);
    content[file_size] = '\0';
    fclose(file);

    const char *html_template = 
        "<!DOCTYPE html>"
        "<html lang=\"en\">"
        "<head>"
            "<meta charset=\"UTF-8\">"
            "<meta name=\"viewport\" content=\"width=device-width, initial-scale=1.0\">"
            "<title>%s</title>"
            "<style>"
                "body { font-family: Arial, sans-serif; line-height: 1.6; padding: 20px; }"
                "pre { background-color: #f4f4f4; padding: 15px; border-radius: 5px; white-space: pre-wrap; word-wrap: break-word; }"
            "</style>"
        "</head>"
        "<body>"
            "<h1>%s</h1>"
            "<pre>%s</pre>"
        "</body>"
        "</html>";

    char *filename = strrchr(path, '/');
    if (filename == NULL) filename = (char *)path;
    else filename++;

    char *html_content = malloc(strlen(html_template) + strlen(filename) * 2 + strlen(content) + 1);
    sprintf(html_content, html_template, filename, filename, content);

    mg_http_reply(c, 200, "Content-Type: text/html; charset=utf-8\r\n", html_content);

    free(content);
    free(html_content);
}

static void ev_handler(struct mg_connection *c, int ev, void *ev_data, void *fn_data) {
    if (ev == MG_EV_HTTP_MSG) {
        struct mg_http_message *hm = (struct mg_http_message *) ev_data;
        
        if (mg_http_match_uri(hm, "/")) {
            check_and_generate_report();
            mg_http_serve_dir(c, hm, s_web_directory);
        } else if (mg_match(hm->uri, mg_str("*.txt"), NULL)) {
            char path[256];
            mg_snprintf(path, sizeof(path), "%s%.*s", s_web_directory, (int) hm->uri.len, hm->uri.ptr);
            serve_text_file(c, path);
        } else {
            mg_http_serve_dir(c, hm, s_web_directory);
        }
    }
}

int main(int argc, char *argv[]) {
    struct mg_mgr mgr;
    struct mg_connection *c;

    mg_mgr_init(&mgr);
    c = mg_http_listen(&mgr, s_http_port, ev_handler, NULL);
    if (c == NULL) {
        printf("Error starting server on port %s\n", s_http_port);
        return 1;
    }

    printf("Starting Mongoose web server on port %s\n", s_http_port);
    for (;;) mg_mgr_poll(&mgr, 1000);
    mg_mgr_free(&mgr);

    return 0;
}

File: draft/start_server.rs
use std::fs;
use std::io::{self, Read};
use std::net::TcpListener;
use std::path::{Path, PathBuf};
use std::process::Command;
use std::time::{Duration, SystemTime};
use tiny_http::{Server, Response, Header};
use html_escape::encode_text;
use clap::Parser;

#[derive(Parser, Debug)]
#[command(author, version, about, long_about = None)]
struct Args {
    #[arg(short, long, default_value_t = 8000)]
    port: u16,

    #[arg(short, long, default_value = ".")]
    directory: String,
}

fn main() -> io::Result<()> {
    let args = Args::parse();
    
    std::env::set_current_dir(&args.directory)?;
    
    let server = Server::http(format!("0.0.0.0:{}", args.port)).unwrap();
    println!("Server running on http://0.0.0.0:{}/", args.port);

    for request in server.incoming_requests() {
        let url = request.url().to_string();
        
        if url == "/" {
            check_and_generate_report();
            serve_file(request, "index.html")?;
        } else if url.ends_with(".txt") {
            serve_text_file(request, &url[1..])?;
        } else {
            serve_file(request, &url[1..])?;
        }
    }

    Ok(())
}

fn check_and_generate_report() {
    let html_file = Path::new("index.html");
    if !html_file.exists() || is_file_older_than_60_seconds(html_file) {
        println!("index.html is older than 60 seconds or does not exist. Running generate_report.py...");
        Command::new("python")
            .arg("generate_report.py")
            .status()
            .expect("Failed to execute generate_report.py");
    } else {
        println!("index.html is up-to-date.");
    }
}

fn is_file_older_than_60_seconds(path: &Path) -> bool {
    path.metadata()
        .and_then(|metadata| metadata.modified())
        .map(|modified| {
            SystemTime::now()
                .duration_since(modified)
                .unwrap_or(Duration::from_secs(0))
                > Duration::from_secs(60)
        })
        .unwrap_or(true)
}

fn serve_file(request: tiny_http::Request, filename: &str) -> io::Result<()> {
    let path = Path::new(filename);
    let content = fs::read(path)?;
    let response = Response::from_data(content);
    request.respond(response).map_err(|e| io::Error::new(io::ErrorKind::Other, e))
}

fn serve_text_file(request: tiny_http::Request, filename: &str) -> io::Result<()> {
    let path = Path::new(filename);
    let mut content = String::new();
    fs::File::open(path)?.read_to_string(&mut content)?;

    let html_content = format!(
        r#"
        <!DOCTYPE html>
        <html lang="en">
        <head>
            <meta charset="UTF-8">
            <meta name="viewport" content="width=device-width, initial-scale=1.0">
            <title>{}</title>
            <style>
                body {{ font-family: Arial, sans-serif; line-height: 1.6; padding: 20px; }}
                pre {{ background-color: #f4f4f4; padding: 15px; border-radius: 5px; white-space: pre-wrap; word-wrap: break-word; }}
            </style>
        </head>
        <body>
            <h1>{}</h1>
            <pre>{}</pre>
        </body>
        </html>
        "#,
        path.file_name().unwrap().to_str().unwrap(),
        path.file_name().unwrap().to_str().unwrap(),
        encode_text(&content)
    );

    let response = Response::from_string(html_content)
        .with_header(Header::from_bytes(&b"Content-Type"[..], &b"text/html; charset=utf-8"[..]).unwrap());

    request.respond(response).map_err(|e| io::Error::new(io::ErrorKind::Other, e))
}
